File: server/typescript/packages/codegen-ts/test/golden/enum-conformance.test.ts
// Cross-port enum conformance (TS). Loads the shared fixture
// `fixtures/codegen-conformance/enum/input/meta.enum.json` and asserts the TS entity
// codegen represents the entity's enum fields idiomatically:
//   • `status` (inline @values) → an `<Entity><Field>` union + inline z.enum, with the
//     DB CHECK constraint (TS owns schema, ADR-0015).
//   • `priority` (extends the root abstract `Priority` enum) → materialized once in the
//     shared `enums.ts` module and referenced by the entity (no inline redeclaration).

import { describe, test, expect } from "bun:test";
import { mkdtempSync, rmSync, readFileSync, readdirSync, existsSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, dirname, resolve } from "node:path";
import { runGen, defineConfig } from "../../src/index.js";
import { entityFile, queriesFile, barrel } from "../../src/generators/index.js";
import { MetaDataLoader, InMemoryStringSource, type MetaRoot } from "@metaobjectsdev/metadata";

function findRepoRoot(start: string): string {
  let dir = start;
  while (true) {
    if (existsSync(join(dir, "fixtures")) && existsSync(join(dir, "server"))) return dir;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("could not locate repo root from " + start);
    dir = parent;
  }
}

async function loadSharedFixture(): Promise<MetaRoot> {
  const fxDir = join(findRepoRoot(import.meta.dir), "fixtures", "codegen-conformance", "enum", "input");
  const sources = readdirSync(fxDir)
    .filter((f) => f.endsWith(".json"))
    .map((f) => new InMemoryStringSource(readFileSync(join(fxDir, f), "utf-8"), { id: f, format: "json" }));
  const res = await new MetaDataLoader().load(sources);
  expect(res.errors, "shared enum fixture must load cleanly").toEqual([]);
  return res.root;
}

async function gen(root: MetaRoot): Promise<Record<string, string>> {
  const tmp = mkdtempSync(join(tmpdir(), "enum-conf-"));
  try {
    await runGen({
      config: defineConfig({
        outDir: tmp,
        extStyle: "none",
        dbImport: "~/db",
        dialect: "postgres",
        generators: [entityFile(), queriesFile(), barrel()],
      }),
      metadata: root,
    });
    const files: Record<string, string> = {};
    for (const name of readdirSync(tmp)) {
      if (name.endsWith(".ts")) files[name] = readFileSync(join(tmp, name), "utf-8");
    }
    return files;
  } finally {
    rmSync(tmp, { recursive: true, force: true });
  }
}

describe("enum codegen-conformance (TS)", () => {
  test("inline enum field emits an <Entity><Field> union + inline z.enum + DB CHECK", async () => {
    const files = await gen(await loadSharedFixture());
    const t = files["Ticket.ts"]!;
    expect(t).toBeDefined();
    expect(t).toContain('export type TicketStatus = "OPEN" | "PENDING" | "CLOSED";');
    expect(t).toContain('z.enum(["OPEN", "PENDING", "CLOSED"])');
    // TS owns the schema → enum CHECK constraint on the column.
    expect(t).toContain("status IN ('OPEN', 'PENDING', 'CLOSED')");
    expect(t).toContain("priority IN ('LOW', 'MEDIUM', 'HIGH')");
  });

  test("extends an abstract root enum → materialized once in shared enums.ts, referenced by the entity", async () => {
    const files = await gen(await loadSharedFixture());
    const enums = files["enums.ts"]!;
    expect(enums, "shared enums.ts must be emitted for the extended abstract enum").toBeDefined();
    expect(enums).toContain('export type Priority = "LOW" | "MEDIUM" | "HIGH";');
    expect(enums).toContain('z.enum(["LOW", "MEDIUM", "HIGH"])');

    const t = files["Ticket.ts"]!;
    // The entity references the shared enum, never redeclaring the union inline.
    expect(t).toContain('from "./enums"');
    expect(t).toContain("PriorityEnum");
    expect(t).not.toContain('export type Priority =');
  });
});

File: server/java/codegen-spring/src/test/java/com/metaobjects/generator/spring/EnumConformanceTest.java
package com.metaobjects.generator.spring;

import com.metaobjects.generator.direct.MultiFileDirectGeneratorBase;
import com.metaobjects.loader.MetaDataLoader;
import com.metaobjects.object.MetaObject;
import com.metaobjects.registry.SharedRegistryTestBase;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertTrue;

/**
 * Cross-port enum conformance (Java/Spring). Loads the shared fixture
 * {@code fixtures/codegen-conformance/enum/input/meta.enum.json} and asserts the entity DTO
 * represents enum fields as a value-constrained Java {@code enum} (parity with TS / Python /
 * Kotlin / C#): an INLINE enum ({@code status}) emits a nested {@code public enum
 * <Entity><Field>} in the record, while a field that {@code extends} the package-level abstract
 * root {@code Priority} enum is materialized ONCE as a standalone top-level {@code Priority.java}
 * (FR-019, ADR-0026) and merely REFERENCED by the DTO (not redeclared inline), with the DTO
 * components typed as those enums rather than {@code String}.
 */
public class EnumConformanceTest extends SharedRegistryTestBase {

    @Rule
    public TemporaryFolder tmp = new TemporaryFolder();

    private MetaDataLoader loadShared() throws IOException {
        Path repoRoot = Path.of(System.getProperty("user.dir")).resolve("../../..").normalize();
        Path fixture = repoRoot.resolve("fixtures/codegen-conformance/enum/input/meta.enum.json");
        assertTrue("shared enum fixture must exist at " + fixture, Files.exists(fixture));
        return SpringTestFixtures.loadFixture(tmp.newFolder().toPath(), "metaEnum", Files.readString(fixture));
    }

    private Path run(MultiFileDirectGeneratorBase<MetaObject> gen, MetaDataLoader loader) throws IOException {
        Path out = tmp.newFolder().toPath();
        Map<String, String> args = new HashMap<>();
        args.put("outputDir", out.toString());
        gen.setArgs(args);
        gen.execute(loader);
        return out;
    }

    @Test
    public void dtoTypesEnumFieldsAsValueConstrainedJavaEnums() throws Exception {
        Path out = run(new SpringDtoGenerator(), loadShared());
        Path dto = out.resolve("acme/shop/TicketDto.java");
        assertTrue("TicketDto.java must be emitted", Files.exists(dto));
        String src = Files.readString(dto);

        // Inline enum: nested `public enum TicketStatus { OPEN, PENDING, CLOSED }` + typed component.
        assertTrue("inline enum decl must be emitted; saw:\n" + src,
            src.contains("public enum TicketStatus { OPEN, PENDING, CLOSED }"));
        assertTrue("status component must be typed as the enum, not String; saw:\n" + src,
            src.contains("TicketStatus status"));

        // FR-019: extends a package-level abstract Priority → materialized standalone, referenced
        // (NOT nested) by the DTO; component still typed as the shared enum.
        assertTrue("shared enum must NOT be nested in the DTO (it is materialized standalone); saw:\n" + src,
            !src.contains("public enum Priority"));
        assertTrue("priority component must be typed as the shared enum; saw:\n" + src,
            src.contains("Priority priority"));

        // The standalone materialized enum file is emitted ONCE in the declaring package.
        Path sharedEnum = out.resolve("acme/shop/Priority.java");
        assertTrue("standalone Priority.java must be materialized; expected at " + sharedEnum,
            Files.exists(sharedEnum));
        String enumSrc = Files.readString(sharedEnum);
        assertTrue("standalone enum must declare the members verbatim; saw:\n" + enumSrc,
            enumSrc.contains("public enum Priority { LOW, MEDIUM, HIGH }"));

        // The enum value set must NOT collapse back to a String component.
        assertTrue("status must not be a String component; saw:\n" + src,
            !src.contains("String status"));
    }
}
